### PR TITLE
Push SessionState into FileFormat (#4349)

### PR DESCRIPTION
**Pull request by tustvold**
# Which issue does this PR close?

Part of #4349
Relates to #4617 
Relates to #3909 #3886

# Rationale for this change

#4427 pushed `ConfigOptions` into `ParquetFormat`, however, the resulting `Arc<Mutex<ConfigOptions>>` is tricky to work with (#3886) and has unclear mutation semantics (#4617). 

# What changes are included in this PR?

Pushes `SessionState` into `FileFormat` which in turn will allow removing `ConfigOptions` from `ParquetFormat` and instead just fetching the configuration from the passed in `SessionState`

# Are these changes tested?

# Are there any user-facing changes?

### CODE REVIEW

File: datafusion/core/src/physical_plan/file_format/avro.rs [hunk 1]
```diff
@@ -254,7 +258,7 @@ mod tests {
             projection: Some(vec![0, 1, 2]),
             limit: None,
             table_partition_cols: vec![],
-            config_options: ConfigOptions::new().into_shareable(),
+            config_options: ctx.config_options(),
```
Comment (tustvold) [inline datafusion/core/src/physical_plan/file_format/avro.rs:261]: I fully intend to remove this in a future PR

File: datafusion/core/tests/row.rs [hunk 2]
```diff
@@ -84,7 +76,7 @@ async fn test_with_parquet_word_aligned() -> Result<()> {
 }
 
 async fn get_exec(
-    ctx: &SessionContext,
+    ctx: &SessionState,
```
Comment (tustvold) [inline datafusion/core/tests/row.rs:79]: Taking `SessionState` here is more correct as it allows the query to be executed against the same state

File: datafusion/core/src/datasource/file_format/avro.rs [hunk 3]
```diff
@@ -47,6 +48,7 @@ impl FileFormat for AvroFormat {
 
     async fn infer_schema(
         &self,
+        _ctx: &SessionState,
```
Comment (alamb) [inline datafusion/core/src/datasource/file_format/avro.rs:51]: I agree this makes much more sense 👍

File: datafusion/core/src/datasource/file_format/avro.rs [hunk 4]
```diff
@@ -124,9 +129,10 @@ mod tests {
     #[tokio::test]
     async fn read_limit() -> Result<()> {
         let session_ctx = SessionContext::new();
-        let task_ctx = session_ctx.task_ctx();
+        let ctx = session_ctx.state();
```
Comment (alamb) [inline datafusion/core/src/datasource/file_format/avro.rs:132]: maybe `state` would be a better name than `ctx` as there is already `session_ctx` which is a Context 🤔

File: datafusion/core/src/datasource/file_format/parquet.rs [hunk 5]
```diff
@@ -1257,13 +1265,13 @@ mod tests {
     }
 
     async fn get_exec(
-        ctx: &SessionContext,
+        ctx: &SessionState,
```
Comment (alamb) [inline datafusion/core/src/datasource/file_format/parquet.rs:1268]: I find the `ctx` name confusing

**Review by alamb — APPROVED** [review]
Thanks -- looks good to me, thank you @tustvold 

I have only a minor quibble about the name used for variables of `SessionState` type (`ctx`) being confusing. Otherwise 👍 

cc @mingmwang  @andygrove

File: datafusion/core/src/datasource/file_format/parquet.rs (hunk 5 above)
Comment (tustvold) [reply datafusion/core/src/datasource/file_format/parquet.rs]: Yeah, I was just copying the naming used in things like TableProvider - I don't disagree that it is confusing. Renamed

**Comment by tustvold** [comment]
I plan to merge this once CI turns green, so that I can base some further changes off it

**Pull request merged**